### ISSUE TITLE
Evaluate adding symlink support

### ISSUE THREAD

**Issue by AdamSvetec** [bufbuild/buf#104]
Hi, another question for you.

Is it possible to have buf follow symlinks? The way the build folder is setup with bazel (by default) means the original repo is symlinked. We need to use the build folder as bazel pulls in all the external proto files we need to compile with.

Thanks,
Adam

**Comment by bufdev** [comment]
We will have to evaluate that, will get back to you.

**Comment by AdamSvetec** [comment]
:+1:  thanks

**Comment by bufdev** [comment]
Do you think you could go through the use case a bit more? If it’s private, feel free to contact us at the email listed in the context section of our website.

I know Bazel includes several symlink directories in the workspace, but they’re just for the user’s convenience (you likely know this, but https://docs.bazel.build/versions/master/output_directories.html )

Also, could this issue be relate at all per chance? https://github.com/bazelbuild/bazel/issues/7057 let us know!

**Comment by bufdev** [comment]
Accidental close - keeping issue open, need to evaluate security concerns of allowing this outside of CLI context

**Comment by AdamSvetec** [comment]
Hey,

So our use case does indeed involve bazel. So we currently pull in a number of external deps through bazel that we import in our own proto files.

So to run buf and have it be able to actually include the external proto files, we need to run it from the "execution_root" folder instead of the source repo because bazel doesn't pull external deps into the actually source repo (which I think is the same as others like conan etc)

By default bazel symlinks the source dir in the execution_root to the actual source dir (ie `ll $(bazel info execution_root)`). Now you're right that you can probably disable this so it does an actually copy whenever you build, but a lot of people find this annoying especially when testing changes to non-compiled language like python where you don't really need to "build" anything to test changes.

So in short, we can't run buf directly in our source repo because external deps aren't pulled there. Bazel sets up something that looks like the "final source dir" with all external deps included that we would like to use to run your tool, but because bazel symlinks the source directory (and other local filesystem repos), this doesn't work with your tool.

Thanks,
Adam

**Comment by AdamSvetec** [comment]
Looked at the docs you sent over a little deeper. Yes the symlinks they refer to are just for convenience (and I now realize we could also use them to build single root dir with all proto files under it). However because we need a single "root" of the repo, either directory we consider the root will still have symlinks involved in order to have source and external proto files under the same directory.

I can make a diagram if it helps, again probably a way around this, but it would create some headaches for us.

Thanks

**Comment by bufdev** [comment]
When you say single root, are you talking about in terms of buf's roots, or your own concept? If in terms of Buf, you can have multiple, but I'm assuming you've figured that part out by now :-)

**Comment by bufdev** [comment]
The tldr is that the lack of symlink support isn't due to any technical limitation, its a limitation we purposefully added due to some things we're working on (we don't want to have to account for symlinks for a few security-related reasons). It's not hard to solve on a technical level, but we want to make sure we're evaluating this properly so we don't cause issues down the road.

I am interested if we can solve it without symlinks, hence the questions :-) There's others who have had some similar issues (not symlinks, but with external deps). Buf is going to have a built-in solution for external remote deps in the coming month or two as well.

**Comment by bufdev** [comment]
You've likely seen this by now, but https://buf.build/docs/build-configuration#roots (and I'm betting this isn't what you're referring to when you say roots in this context)

**Comment by AdamSvetec** [comment]
Sorry yes I'm using the word root in the wrong way likely.

Pulling the terminology from the page you linked:
"""
roots#
This is a list of the directories that contain your .proto files. The directory paths must be relative to the root of your repository, and cannot point to a location outside of your repository. They also represent the root of your import paths within your .proto files.
"""
The issue is just that external deps get pulled to a location that IS outside of the repository root (by bazel). So no way to list those external deps as a second root. maybe I'm missing something

**Comment by bufdev** [comment]
So just to make sure (I used to use bazel, but have forgotten a bit here heh), roughly, your situation is:

- `/home/adam/dev/bazel_repo/proto` has your source .proto files
- `/home/.cache/where_bazel_caches_reps/external/repo` has some external deps

Is this an accurate assessment?

**Comment by AdamSvetec** [comment]
yes

**Comment by AdamSvetec** [comment]
Seemingly I guess I could use /home/adam as the repo root? but that seems wrong no?

**Comment by bufdev** [comment]
I was about to comment. Technically you could do that yes, with `dev/bazel_repo/proto` and `.cache/bazel/external/repo` as the configured Buf roots.

To give some context, the goal with Buf is to have all Protobuf files derive from a common directory, usually your repository root - think of it like Dockerfiles, the goal is for all context for Buf to be relative to a common directory, (you can probably figure out why via https://buf.build/docs/roadmap).

This doesn't mean we shouldn't cover more use cases though. Supporting Bazel repositories as first-class citizens is a goal of ours, so this is an important case we want to cover.

For now, you can get around this by doing `buf check lint --input /home/adam --input-config '{"build":{"roots":["dev/bazel/proto", ".cache/bazel/external/repo"]}}'`, that should work out of the box, but we'll work on this. It might be as simple as enabling symlinks for certain contexts.

**Comment by bufdev** [comment]
https://buf.build/docs/inputs will give some more context on the `--input` flag FYI

**Comment by AdamSvetec** [comment]
gotcha yes makes sense, I appreciate your help. I will follow along to see progress here. would also be happy to contribute if you think that's doable.
I think it would be as simple as a flag like you said. Bazel nicely makes symlinks for you as you pointed out, so from:
/home/adam/dev/bazel_repo/proto
all you would really need to do (if symlinks were supported) is:
buf check lint --input /home/adam/dev/bazel_repo/proto --input-config '{"build":{"roots":["proto", "bazel-bazel_repo/external/repo"]}}'

Doesn't really save much, but avoids the annoyance of having to figure out common root on the fly.

And I think the git functionality breaks if your --input isn't actually a git repo root right?

**Comment by bufdev** [comment]
Not sure what you're referring to :-) Can you elaborate (re git functionality)?

**Comment by AdamSvetec** [comment]
the --against-input git functionality in buf check breaking

give me one sec to make an example

**Comment by bufdev** [comment]
Ah - so interesting question. Yes, you can't do `--against-input /home/adam` as an input needs to be a self-contained git repository that Buf can compile and then compare your current files to - one of the core arguments about why we want files to be self-contained in a logical directory (or tarball, or git repository). The breaking change detector works by considering different inputs to be completely logically separate - it compiles the main input, then compiles the against input compiletely independently, and compares the two results.

I think the bigger overarching issue here is having a good way to deal with external dependencies, or handling symlinks. The former is on the way, the latter we need to consider a bit more in terms of the CLI - allowing symlinks actually hurts the 99% case for other users of Buf, and can lead to some ugly results, so we need to see if there's a good way to do it (perhaps by an opt-in flag).

**Comment by plaird** [comment]
@AdamSvetec I am about to embark on implementing _buf_ for our Bazel workspace, and found this thread. Any chance you have published (or are willing to) your Bazel toolchain defs for this?

**Comment by AdamSvetec** [comment]
Hey @plaird 
I actually ended up using protolock because it's able to parse proto tree without needing external deps :(
The plan is definitely to switch back once this is better supported/I have more time to figure this out. I wasn't familiar enough with bazel to make that effort worth it. If you do manage to make this work, please let me know or comment on this thread with your solution as it'd be super helpful.
Thanks!

**Comment by cgthayer** [comment]
Use case.  We're building against a bunch of external protobufs which get pulled into git via "submodules" then we symlink into the right places to create a nice tree for proper "import"ing.  We could do something similar with via "proto_roots" but this setup 
works with a bunch of tooling consistently at the filesystem level.

We don't want to simply use "roots: ." because that pulls in everything (twice). And pointing at the symlink directories results in "no .proto target files found"

```
::: SM-001907CT 02:52:14 (cgt-initial) 0 protos; ll                                                                                                                                                                            uth/service
total 320
drwxrwxrwx 1 thayer thayer    512 Oct 21 14:52 ./                                                                                                                                                                               ice-auth/se
drwxrwxrwx 1 thayer thayer    512 Oct 21 10:30 ../
-rwxrwxrwx 1 thayer thayer    168 Sep 24 23:30 README.md*                                                                                                                                                                       ed. [/c/src
-rwxrwxrwx 1 thayer thayer    113 Oct 21 14:06 buf.gen.yaml*
-rwxrwxrwx 1 thayer thayer    275 Oct 21 14:49 buf.yaml*                                                                                                                                                                        ce-auth/ser
drwxrwxrwx 1 thayer thayer    512 Sep 24 23:30 echo-grpc/
lrwxrwxrwx 1 thayer thayer     24 Oct 15 15:22 envoy -> external-envoy/api/envoy/                                                                                                                                               /envoy-auth
drwxrwxrwx 1 thayer thayer    512 Oct 13 21:25 envoy-auth/
-rwxrwxrwx 1 thayer thayer  14311 Oct 13 20:43 ext_authz.proto*                                                                                                                                                                 h/src/Envoy
drwxrwxrwx 1 thayer thayer    512 Oct 15 15:19 external-envoy/
drwxrwxrwx 1 thayer thayer    512 Oct 14 13:47 external-google/                                                                                                                                                                 src/Roblox/
drwxrwxrwx 1 thayer thayer    512 Oct 20 11:42 external-opencensus/
drwxrwxrwx 1 thayer thayer    512 Oct 20 12:11 external-openmetrics/                                                                                                                                                            ox/bedev2-s
drwxrwxrwx 1 thayer thayer    512 Oct 20 12:17 external-prometheus/
drwxrwxrwx 1 thayer thayer    512 Oct 14 21:48 external-protoc-gen-validate/                                                                                                                                                    Roblox/bede
drwxrwxrwx 1 thayer thayer    512 Oct 14 13:55 external-udpa/
lrwxrwxrwx 1 thayer thayer     22 Oct 15 15:23 google -> external-google/google/                                                                                                                                                ev2-service
-rwxrwxrwx 1 thayer thayer 311094 Oct 21 14:38 ls.out*
lrwxrwxrwx 1 thayer thayer     33 Oct 20 12:17 metrics.proto -> external-prometheus/metrics.proto*                                                                                                                              /envoy-auth
lrwxrwxrwx 1 thayer thayer     34 Oct 20 11:43 opencensus -> external-opencensus/src/opencensus/
drwxrwxrwx 1 thayer thayer    512 Sep  1 09:53 roblox/                                                                                                                                                                          ox/bedev2-s
lrwxrwxrwx 1 thayer thayer     19 Oct 20 12:49 test -> external-envoy/test/
lrwxrwxrwx 1 thayer thayer     22 Oct 20 14:49 tools -> ./external-envoy/tools/                                                                                                                                                 uth/service
lrwxrwxrwx 1 thayer thayer     18 Oct 15 15:23 udpa -> external-udpa/udpa/
lrwxrwxrwx 1 thayer thayer     37 Oct 15 15:25 validate -> external-protoc-gen-validate/validate/                                                                                                                               voy-auth/sr
:::: SM-001907CT 02:52:16 (cgt-initial) 0 protos; git submodules status
git: 'submodules' is not a git command. See 'git --help'.                                                                                                                                                                       uth/src/Env

The most similar command is                                                                                                                                                                                                     ces/envoy-a
        submodule
:::: SM-001907CT 02:52:23 (cgt-initial) 0 protos; git submodule status                                                                                                                                                          voy-auth/sr
 c318156496edc46c844822a3d1f107ee496fa449 external-envoy (v1.15.0-798-gc31815649)
 215c12ade72d9d9616457d9b8b2f8a37f38e79f3 external-google (common-protos-1_3_1-2730-g215c12ade)                                                                                                                                 /EnvoyAuth.
 54c9a4c53e1fa422b92e3ae1cd6c72e181f744db external-opencensus (v0.2.1-8-g54c9a4c)
 120af2ae8ff9b61f9d93668380314c3bd41e4270 external-openmetrics (heads/master)                                                                                                                                                   -auth/servi
 60555c9708c786597e6b07bf846d0dc5c2a46f54 external-prometheus (v0.2.0-2-g60555c9)
 ec9cd95372b9630aafba2cb4a0e448c43768e0df external-protoc-gen-validate (v0.3.0-java-26-gec9cd95)                                                                                                                                ices/envoy-
 7e6fe0510fb53d1053138b61476ec271b519602c external-udpa (v0.0.1-3-g7e6fe05)
:::: SM-001907CT 02:52:30 (cgt-initial) 0 protos;                                                                                                                                                                               vice-auth/s
:::: SM-001907CT 02:52:35 (cgt-initial) 0 protos; cat buf.yaml
version: v1beta1                                                                                                                                                                                                                h/src/Envoy
build:
  roots:
    - envoy/.

lint:                                                                                                                                                                                                                           o3";' to sp
  use:
    - BASIC
    - FILE_LOWER_SNAKE_CASE
  except:
    - ENUM_NO_ALLOW_ALIAS
    - IMPORT_NO_PUBLIC
    - PACKAGE_AFFINITY
    - PACKAGE_DIRECTORY_MATCH
    - PACKAGE_SAME_DIRECTORY
breaking:
  use:
    - WIRE_JSON
:::: SM-001907CT 02:52:39 (cgt-initial) 0 protos; buf ls-files
no .proto target files found
```

**Comment by bufdev** [comment]
We're looking at if there is a path forward here - our primary concern is security, as the same code in the Buf CLI is also used in the backend - we're concerned with what symlinks enable in that context, and providing a different capability from the Buf CLI as opposed to our upcoming Buf Schema Registry can cause some issues down the road. We're actively investigating this though.

**Comment by bufdev** [comment]
https://github.com/bufbuild/buf/pull/198 implements the first part of this - we'll have a follow-up PR that does the rest.

**Comment by bufdev** [comment]
@AdamSvetec @plaird @cgthayer can you do `go get github.com/bufbuild/buf/cmd/buf@master` and see if this now works for you if you have time? We've added symlink support, this will go out in the next release. Would be good to make sure this covers your use cases though.

**Comment by bufdev** [comment]
This is released as v0.32.0 Let us know if you have any issues!

**Comment by alokpr** [comment]
Is there a way to make buf ignore the symlinks? It is picking up convenience symlinks created by bazel. Before v1 I was able to use the `roots` feature to explicitly specify the directory that contained all my protos. But now I need to specify `excludes` which is not possible because convenience symlinks do not have a fixed name. They depend on how the user has configured bazel.

**Comment by bufdev** [comment]
Fair request, see https://github.com/bufbuild/buf/issues/908
